### PR TITLE
Feat/ci enhancements

### PR DESCRIPTION
**Pull request by samurato**
## Summary by CodeRabbit

- **New Features**
  - Deployment process now triggers on both push and pull request events.
  - Added a new step to update pull requests with Terraform plan output in a collapsible format.

- **Improvements**
  - Enhanced clarity of Terraform command outputs by adding `-no-color` flag to various tasks.
  - Indentation corrections for better structure in the task definitions.

### CODE REVIEW

File: .github/workflows/deploy-to-gcp.yaml [hunk 1]
```diff
@@ -1,16 +1,20 @@
 name: Deploy Xplorers API to GCP
-
-on: push
+on: 
+  push:
+  pull_request:
```
Comment (boltdynamics) [inline .github/workflows/deploy-to-gcp.yaml:4]: does this restrict the workflow to pull_requests only? Will this work when we merge to `main`?

And should we be scoping this to just when a Pull Request is ready for review?

```
on:
  pull_request:
    types:
      - ready_for_review
```

File: .github/workflows/deploy-to-gcp.yaml [hunk 2]
```diff
@@ -1,16 +1,20 @@
...
 
 permissions:
-  contents: read
   id-token: write
+  contents: write
+  pull-requests: write
+  repository-projects: write
```
Comment (boltdynamics) [inline .github/workflows/deploy-to-gcp.yaml]: Not sure why we need `repository-projects`? Does this write to the project board?
image

File: .github/workflows/deploy-to-gcp.yaml [hunk 3]
```diff
@@ -1,16 +1,20 @@
...
-
-on: push
+on: 
+  push:
+  pull_request:
 
 permissions:
-  contents: read
   id-token: write
+  contents: write
```
Comment (boltdynamics) [inline .github/workflows/deploy-to-gcp.yaml:9]: Is this to allow Github Actions to create a release?
image

File: .github/workflows/deploy-to-gcp.yaml [hunk 4]
```diff
@@ -37,8 +41,32 @@ jobs:
           sudo snap install task --classic
 
       - name: Terraform Plan
+        if: github.event_name == 'pull_request'
```
Comment (boltdynamics) [inline .github/workflows/deploy-to-gcp.yaml:35]: Do we need to explicitly call out the event_name to be `pull_request` here given we already define this on top of the file?
```
on: 
  push:
  pull_request:
```

File: Taskfile.yml [hunk 5]
```diff
@@ -108,6 +108,7 @@ tasks:
         terraform init \
         -backend-config="bucket=${ARTIFACTS_BUCKET_NAME}" \
         -backend-config="prefix=${ARTIFACTS_BUCKET_TERRAFORM_PREFIX}"
+      - terraform validate -no-color
```
Comment (boltdynamics) [inline Taskfile.yml:111]: Noice 👍 

`-no-color - Disables terminal formatting sequences in the output. Use this if you are running Terraform in a context where its output will be rendered by a system that cannot interpret terminal formatting.`

**Comment by boltdynamics** [comment]
Can we format the output of `Terraform plan`?
image

File: .github/workflows/deploy-to-gcp.yaml (hunk 2 above)
Comment (samurato) [reply .github/workflows/deploy-to-gcp.yaml]: Made changes Thanks

File: .github/workflows/deploy-to-gcp.yaml (hunk 4 above)
Comment (samurato) [reply .github/workflows/deploy-to-gcp.yaml:35]: Applied

File: .github/workflows/deploy-to-gcp.yaml (hunk 3 above)
Comment (samurato) [reply .github/workflows/deploy-to-gcp.yaml:9]: We need this to write PR comments for the action below to print the plan in comments.

**Comment by samurato** [comment]
@boltdynamics  Formatting is fixed.

File: .github/workflows/deploy-to-gcp.yaml [hunk 6]
```diff
@@ -1,43 +1,61 @@
...
+            \`\`\`
+            \n
+            </details>
+            *Pushed by: @${{ github.actor }}, Action: \`${{ github.event_name }}\`*`;
+            github.rest.issues.createComment({
+              issue_number: context.issue.number,
+              owner: context.repo.owner,
+              repo: context.repo.repo,
+              body: output
+            })
```
[the remark this answers is not in the document]
Comment (samurato) [reply .github/workflows/deploy-to-gcp.yaml:60]: False positive

File: .github/workflows/deploy-to-gcp.yaml (hunk 1 above)
Comment (samurato) [reply .github/workflows/deploy-to-gcp.yaml:4]: This should do the trick as its looking for a push event in branch main. However Suhail dai has suggested that we do the deployment (Terraform Apply) manually when tthe TF files are approved and someone do it.
So doing some modifications.

**Review by boltdynamics — APPROVED** [review]
Thanks for making the suggested changes 🚀

**Pull request merged**